### ISSUE TITLE
Issues found on new version wip-invisible

### ISSUE THREAD

**Issue by ProxiBlue** [ProxiBlue/reCaptcha#23]
This relates to new version in branch : https://github.com/ProxiBlue/reCaptcha/tree/wip-invisible

**Comment by vishy93** [comment]
@ProxiBlue First issue, when hovering over the collapsing "protected by reCaptcha", and you move your mouse over to the Blue rectangle it disappears quickly,  and thus making the ability to click the privacy and terms none.

**Comment by vishy93** [comment]
Second issue, it is affixed without choice to bottom right, and it might affect design. I.e, if you have a back to top link in bottom right.

**Comment by ProxiBlue** [comment]
@vishy93  To trigger reCaotcha, set your browser identifier to 'googlebot'

https://winaero.com/blog/change-user-agent-chrome/

it will immediately trigger images reCaptcha

**Comment by ProxiBlue** [comment]
@vishy93 

Updated to allow position setting, using admin settings.

**Comment by ProxiBlue** [comment]
@vishy93 

> First issue, when hovering over the collapsing "protected by reCaptcha", and you move your mouse over to the Blue rectangle it disappears quickly, and thus making the ability to click the privacy and terms none.

I cannot reproduce. Can you confirm same issue is with base RWD theme, or is it only with your own theme.

**Comment by vishy93** [comment]
Base RWD theme on latest chrome browser

**Comment by ProxiBlue** [comment]
https://drive.google.com/file/d/1vmk3wygoB335qnUdVVC2qVOi5aBCAsAa/view

Possible a browser compatibility issue. I am using Chrome 68

**Comment by vishy93** [comment]
Works well there.

**Comment by ProxiBlue** [comment]
I have determined is not working on checkout. I am investigating why.

**Comment by vishy93** [comment]
@ProxiBlue I've not tried in checkout but will have a look.

**Comment by vishy93** [comment]
@ProxiBlue not sure if you have noticed, but thought it might be useful, when you login on checkout for a brief moment you can see that it correctly logins you in and takes you to checkout but then it literally after like half a second or so redirects you back and sets the "Incorrect CAPTCHA." error.

**Comment by vishy93** [comment]
@ProxiBlue Also if your logged in is intentionally not meant to show in the contacts page?

**Comment by ProxiBlue** [comment]
Updated wip-invisible branch with code that makes checkout work.

So far I have tested: Login, create account, checkout login, checkout as guest, checkout with register

**Comment by ProxiBlue** [comment]
@vishy93 

> Also if your logged in is intentionally not meant to show in the contacts page?

Can you elaborate? I am not sure what you mean

**Comment by ProxiBlue** [comment]
@vishy93 

> not sure if you have noticed, but thought it might be useful, when you login on checkout for a brief moment you can see that it correctly logins you in and takes you to checkout but then it literally after like half a second or so redirects you back and sets the "Incorrect CAPTCHA." error.

Checkout was not working, latest update is adjusted for checkout. The code from the initial PR that got me working on this, broke checkout completely for recaptcha, so I also reverted that.

Can you check with latets code, see if this is still happening.

**Comment by ProxiBlue** [comment]
I have tested all sections that require reCaptcha, and they all seem to be working with invisible

**Comment by vishy93** [comment]
@ProxiBlue I meant if your logged in as a customer and go to contact us the reCapcha won't be there. Only shows on contact us if you are not logged in

**Comment by ProxiBlue** [comment]
This is intentional. If person is logged in, there should be no need to check.

You get spammed via bots on contact us page, without being logged in.

**Comment by ProxiBlue** [comment]
New feature:

https://github.com/ProxiBlue/reCaptcha/tree/wip-invisible#newsletter-subscribe-captcha

**Comment by vishy93** [comment]
Awesome :)

**Comment by vishy93** [comment]
@ProxiBlue Error for Newsletter Subscriptions:

`

Uncaught Error: grecaptcha.execute only takes the 'action' parameter
  | Qs | @ | recaptcha__en_gb.js:417 | (anonymous) | @ | (index):987

`

Also, downloading the new zip only get the app folder, so no skin or js folder.

**Comment by vishy93** [comment]
Also,

`Uncaught Error: reCAPTCHA has already been rendered in this element
    at Object.Ms [as render] (VM21174 recaptcha__en_gb.js:415)
    at (index):357
    at prototype.js:825
    at Array.forEach (<anonymous>)
    at Array.each (prototype.js:824)
    at window.reCaptchaCallback ((index):350)
    at ws (VM21174 recaptcha__en_gb.js:422)
    at xs (VM21174 recaptcha__en_gb.js:406)
    at VM21174 recaptcha__en_gb.js:423
    at VM21174 recaptcha__en_gb.js:432`

**Comment by ProxiBlue** [comment]
> Also, not sure if this is allowed but I quite like how dropbox have got the privacy and terms message in the sign page https://www.dropbox.com/

They used their own text box, with links, and own styles on an inline display of the reCaptcha element

ref: https://stackoverflow.com/questions/44543157/how-to-hide-the-google-invisible-recaptcha-badge

You can do the same by using the INLINE option in admin.

**Comment by ProxiBlue** [comment]
> Also, downloading the new zip only get the app folder, so no skin or js folder.

That is correct. The skin folder was deleted, as it only contained images for V1, which is now obsolete
I don't think there ever was a js part....

BTW - noticed they are beta testing v3!

**Comment by vishy93** [comment]
@ProxiBlue Indeed I was just more concerned about the "These terms do not grant you the right to use any branding or logos used in our Services. Don’t remove, obscure, or alter any legal notices displayed in or along with our Services."

**Comment by vishy93** [comment]
V3 looks very good. I like the newer approach less user experience hinderence. Did you see the other two errors I found?

**Comment by ProxiBlue** [comment]
Yes. Will be looking at that later

**Comment by vishy93** [comment]
@ProxiBlue The script is being called twice. 


<script
--
  | src="https://www.google.com/recaptcha/api.js?onload=reCaptchaCallback&render=explicit&hl=en-GB"
  | async
  | defer></script>

This may be the issue for the second issue

**Comment by ProxiBlue** [comment]
>Also, If you use on two locations at once, for example contact us page and the newsletter subscription form which is situated in the footer you get the error:

Uncaught Error: reCAPTCHA has already been rendered in this element at Object.Ms [as render] (VM21174 recaptcha__en_gb.js:415) at (index):357 at prototype.js:825 at Array.forEach (<anonymous>) at Array.each (prototype.js:824) at window.reCaptchaCallback ((index):350) at ws (VM21174 recaptcha__en_gb.js:422) at xs (VM21174 recaptcha__en_gb.js:406) at VM21174 recaptcha__en_gb.js:423 at VM21174 recaptcha__en_gb.js:432

screen shot 2018-08-12 at 18 19 19

I cannot reproduce this.

Can you ensure your browser cache  is cleared, and don't have any old/previous recapctha.phtml in theme folders.

https://drive.google.com/file/d/1XwCacjgcVnnULQBta-5X6lLFaUp-TAAh/view

**Comment by ProxiBlue** [comment]
>Uncaught Error: grecaptcha.execute only takes the 'action' parameter
  | Qs | @ | recaptcha__en_gb.js:417 | (anonymous) | @ | (index):987
> To replicate error, set your user-agent as either 'Google smartphone' or 'Android (4.0.2) Browser - Galaxy Nexus'.

I can reproduce this one.

**Comment by vishy93** [comment]
@ProxiBlue I have disabled cacheing :) also I always clear. This is my recaptcha.phtml file

[https://hastebin.com/ragoxofawa.xml](https://hastebin.com/ragoxofawa.xml)

![screen shot 2018-08-13 at 13 52 59](https://user-images.githubusercontent.com/9990517/44032825-3d5a3dce-9f00-11e8-966c-6ed3f4bf0246.png)

**Comment by ProxiBlue** [comment]
Soled issue with error:  Uncaught Error: grecaptcha.execute only takes the 'action' parameter

**Comment by ProxiBlue** [comment]
You are using online option?

**Comment by vishy93** [comment]
@ProxiBlue just testing different things out, the error is there in in-line, bottom-left and bottom-right.

**Comment by vishy93** [comment]
@ProxiBlue I think the below might be a contributing factor:

> @ProxiBlue The script is being called twice.

<script --   | src="https://www.google.com/recaptcha/api.js?onload=reCaptchaCallback&render=explicit&hl=en-GB"   | async   | defer></script>
This may be the issue for the second issue

**Comment by vishy93** [comment]
@ProxiBlue 

> 

Soled issue with error: Uncaught Error: grecaptcha.execute only takes the 'action' parameter
--

Sorry, what was the fix for this issue?

**Comment by ProxiBlue** [comment]
That's the thing, I am not getting that, it is only called once....for me here.

**Comment by ProxiBlue** [comment]
>Soled issue with error: Uncaught Error: grecaptcha.execute only takes the 'action' parameter
Sorry, what was the fix for this issue?

The execute was incorrectly loading elements that was passed. Seems those elements are not working correctly when those mobile devices are loaded/used.

**Comment by vishy93** [comment]
@ProxiBlue What browser would you be using? I am using Chrome and can confirm I am using latest package.

![screen shot 2018-08-13 at 13 58 14](https://user-images.githubusercontent.com/9990517/44033480-6c83604c-9f02-11e8-8f1f-d0cd81704c23.png)

**Comment by ProxiBlue** [comment]
![selection_547](https://user-images.githubusercontent.com/4994260/44033522-351e645e-9f3d-11e8-9e55-95af8e5389ce.png)

**Comment by vishy93** [comment]
> The execute was incorrectly loading elements that was passed. Seems those elements are not working correctly when those mobile devices are loaded/used.

@ProxiBlue Ah okay, how would we fix this?

**Comment by ProxiBlue** [comment]
Should already be fixed in latest code.

**Comment by ProxiBlue** [comment]
trying to reproduce other issue

**Comment by vishy93** [comment]
@ProxiBlue Could you view your page source and do a find search for recaptcha/api. I see two instances.

![screen shot 2018-08-13 at 14 15 22](https://user-images.githubusercontent.com/9990517/44033787-5b1c4c82-9f03-11e8-9938-c6015c7bc839.png)

**Comment by ProxiBlue** [comment]
Chrome itself - one instacne of script loading

![selection_551](https://user-images.githubusercontent.com/4994260/44034324-68c8e0e8-9f3f-11e8-992f-e766416a86be.png)

**Comment by vishy93** [comment]
@ProxiBlue [https://hastebin.com/sonidifeqo.http](url)

**Comment by ProxiBlue** [comment]
The code here:

https://github.com/ProxiBlue/reCaptcha/blob/wip-invisible/app/code/community/ProxiBlue/ReCaptcha/Block/Captcha/Recaptcha.php#L14

is supposed to prevent double code insertion.

It sets a registry flag on first call.

which is called from the template:

https://github.com/ProxiBlue/reCaptcha/blob/wip-invisible/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L8

so if teh templaet 'renders' again, that flag will/shoudl result in false, thus preventing a second instance of the code block

It looks like this is not working for you....

Can you check/possible for you to chekc if those bits of code are being executed? is the flag being set?

**Comment by ProxiBlue** [comment]
Can you show me your subscribe.phtml? as well as /template/contacts/form.phtml

**Comment by ProxiBlue** [comment]
Is your magento cache disabled? or enabled?

**Comment by ProxiBlue** [comment]
magento version?

**Comment by vishy93** [comment]
@ProxiBlue 

> Can you check/possible for you to check if those bits of code are being executed? is the flag being set?

How do I do this?

For the requested files see here:

[Subscribe.phtml](https://hastebin.com/gufahibavu.xml)
[form.phtml](https://hastebin.com/pozokaronu.xml)

Magento version 1.9.3.9

**Comment by ProxiBlue** [comment]
bit stumped. will try this on a different magento install tomorrow, latest 1.9.3.9, from a live site dev install, see if I can reproduce.

for me, it works. also tried firefox

![selection_552](https://user-images.githubusercontent.com/4994260/44035308-c61098ca-9f41-11e8-8e14-37570482ec48.png)
![selection_553](https://user-images.githubusercontent.com/4994260/44035315-c83216d8-9f41-11e8-952c-a8aea54f5fb8.png)

Could be 1.9.3.9.

Will try on that version tomorrow, late here now, need to go relax, and spend time with family.

**Comment by vishy93** [comment]
@ProxiBlue I get same result on Firefox with the error, I will see if I can look into it further. Yeah go enjoy precious time my friend we can pick up again tomorrow.

**Comment by vishy93** [comment]
@ProxiBlue the render check is failing.

`<?php var_dump($this->canRenderJS()); ?>`

This is returning true each time, so magento can't prevent loading the additional times.

**Comment by ProxiBlue** [comment]
Strange. So that means the registry flag is not being persistent which means those renderings are generated accross multiple page requests, or the registry is being cleared (which usually happens per page request) 

Can you check if you have any 404s on the page loading. For anything, but specifically of interest is images or Css or js files that are 404ing.

**Comment by ProxiBlue** [comment]
Doing that today.

**Comment by vishy93** [comment]
@ProxiBlue could it be a cache issue, as the Footer is a cached block?

**Comment by ProxiBlue** [comment]
I have tested this in a 1.9.3.9 site, with and without cache, and I only get one instance. 

The fact that you get a new registry entry (so previous registry flag is not set/active) indicates that these are new page requests. 

Are you testing on a new install, or are there other 3rd party modules installed? 

Can you test on a completely different environment?

**Comment by ProxiBlue** [comment]
I am adding some code that will prevent the code from auctioning again, even if in dom multiple times. 

This circumvent the issue, but does not solve the core oddity.

**Comment by vishy93** [comment]
@ProxiBlue That is so strange, what would cause multiple new page request each time?

I am testing on my local install of a site, there are some 3rd party modules.

I can test on a fresh local install with no 3rd party modules now.

**Comment by vishy93** [comment]
@ProxiBlue Okay, so testing on a fresh new install: version 1.3.9.9; I can rightly see the issue is not there. But in my local install is there. What could be causing the registry key issue?

**Comment by ProxiBlue** [comment]
There can be many reasons. One that comes to mind is that you have 404s (images/css/js) that is not loading which then results into a new page request, new session etc.

Check that your local webserver environment has the required error dosuments set correctly, and don't reload the entire magento frameowrk when it hits a 404 for an asset that is not present.


Today, I integrated this change into a production site, and it will be entering UAT testing. Everything sofar just worked out the box.

It will take a couple of weeks to be tested properly before deploy live, once I have that live, and no issues found, the branch will be merged to master

I have paying work waiting, so I also need to put this aside for a while.

**Comment by vishy93** [comment]
@ProxiBlue I can confirm that there is no 404s have thoroughly checked all that.

For a test I changed the `canRenderJS` function like so:

`    public function canRenderJs()
    {
        if (!Mage::registry('recaptcha_rendered')) {
            Mage::register('recaptcha_rendered', true, true);
            return "The if condition is true";
        }
        return false;
    }`

Interestingly, The first render on the contact page obviously shows this text in the var dump. However, on the second render for the newsletter subscription in footer I get:


BOOL(TRUE)

So, something on the additional render calls is setting this registry to true somehow or using an old value.

BINGO!!! I disabled the cache and can see that it is now showing BOOL(FALSE). It is a cacheing issue, on subsequent renders it uses the cached registry value.

**Comment by vishy93** [comment]
@ProxiBlue I have reproduced the error in fresh install rwd too:

1) clear/delete cache (rm -f /var/cache)

2) Go to frontend, then go to contacts. **!important**

3) You will see the error.

4) if you clear the cache then go direct to the contacts page it works sweet but then it won't render elsewhere after for newsletter.

**Comment by ProxiBlue** [comment]
Nice going in reproducing. Still visible paid project. Should get some time in comming week.

**Comment by vishy93** [comment]
@ProxiBlue hey pal, thats fine. I'll keep digging away at it.

**Comment by ProxiBlue** [comment]
@vishy93 

I can now reproduce the issue without fail. Thank you for working that out.

![selection_589](https://user-images.githubusercontent.com/4994260/44618184-cef37f00-a8a3-11e8-8599-99829a1d404c.png)

**Comment by ProxiBlue** [comment]
@vishy93 

Please check updated code in wip-invisible branch
Note that the way to set the newsletter captcha has now also been changed to work like the others by using: 

```<?php echo $this->getChildHtml('recaptcha'); ?>```

**Comment by vishy93** [comment]
@ProxiBlue I will take a look soon as I can

**Comment by vishy93** [comment]
@ProxiBlue wouldn't the  generateCatchaBlock helper method now be redundant? as I can't see it being used anywhere. Thus, can't we remove it?

**Comment by ProxiBlue** [comment]
Yes and no.

Not using it, but someone might be able to use it where layout XML is not available or suitable.

It does not hurt anything to stay there.

**Comment by ProxiBlue** [comment]
It was not the reason for the issue found. Not sure why I did not just use a layout directive in the first place.

**Comment by vishy93** [comment]
@ProxiBlue That is true. Nevertheless, testing the latest version out, when viewing the page source, for instance on the login page, I can see that the script is included twice.

**Comment by ProxiBlue** [comment]
Ok. Makes sense. The main issue here is cache. The block has rendered now with attributes set which is flagging it as already rendered.

I have an idea. Will adjust.

Some great testing. Appreciated.

**Comment by vishy93** [comment]
@ProxiBlue  yeah it's the dammm footer block causing it. I tried:

`<?php echo $this->getChildHtml('recaptcha', false); ?>`

**Comment by ProxiBlue** [comment]
> What if we add the JavaScript only the once?

Technically it is. The second one is via the cached block (which is the core issue)
The flag to prevent the js block from inseting more than once IS only run once, as the cached version in footer never runs it again.

a bit of a catch 22

**Comment by vishy93** [comment]
Yes, but I mean like scrap the whole flag way and go about it by like for example adding the js via addjs.

**Comment by ProxiBlue** [comment]
The is is dynamic. Depending on settings and configuration parts are included and values are set.

Thus it needs to be generated via a blocks output.

If the is was static, identical for all, then placing it via XML layout would have been possible.

**Comment by ProxiBlue** [comment]
@vishy93 

Another update, cannot reproduce either of the issues found, also tested checkout, which has 3 recaptcha elements (login/guest/register)

Give it a try now.

No ajax.

**Comment by vishy93** [comment]
@ProxiBlue Awesome!!! those previous issues look resolved. However, the same script is still added multiple times for each instance of reCaptcha.

**Comment by ProxiBlue** [comment]
Yes, that is not an issue. The code detects that and only one actions.

This cannot be prevented, due to cache. If a block is rendered, which contains captcha (footer) and you enter a page that also add captcha ( example contact us) there is no way to prevent the already cached block from rendering the scripts. They are cached. Since footer is send to page rendering after the contacts form, at the point of the contacts form rendering, there is no way to detect that the footer may, or may not, contain the code.

Bring in full page cache, and the issue gets even more complicated.

Multiple instances can exist now, as the code therein deals with this correctly.

**Comment by ProxiBlue** [comment]
In the same way, if the footer does not include the code (as it was previously rendered in body) moving to a page with no other captcha results in no code in footer, as it is cached, and won't render again. (The issue you foundl

**Comment by vishy93** [comment]
@ProxiBlue I understand that completely. Shame because of cacheing we have to include the script each time.

**Comment by vishy93** [comment]
@ProxiBlue that is great news, Once I go live with it I shall also keep you posted if any issues found. Most likely none :)
